### ISSUE TITLE
Bug: [no-unsafe-enum-comparison] should report when union is in switch and cases have union type

### ISSUE THREAD

**Issue by Afgan0r** [typescript-eslint/typescript-eslint#11596]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.9.2&showAST=types&fileType=.tsx&code=KYOwrgtgBAIgggFTgfQQTQAoFEoG8BQUUAYgJYBOAzgC5QC8UADADSFQDKwAxgPYgAm9KAEZWAX3z5qATwAOwKAHFQwcgENqwfgjkKGBIjPkAuJlAA%2BIgNz4J%2BXiBpRywSrL6Vgp5SFUatOvJCBlBGXiLikpQA7qTUXAAWUAAULm4ewAB0YQCUeGxcap6wiCjo2JlkVNTGBUUK8EiomFiZnA78tUT8wABmamAANjW2kkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2CACYANFu1Zt2SMPjwA9hwAiiAGaj4%2BAMLDkiAGKXoAUSGi4lKIAMpcPLxq%2BHCIBiamZJZMyJQAJojQdo4I%2BP4iYhLS3tAAqkyUSagYjvAecfFQ0IgAjrCUTVlO%2BMUAckllFUxqNR6G2AC%2BYwC6WpPjQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
enum DATA_TYPE {
  First = 0,
  Second = 1,
}

type GeneratedType = {
  type: 0 | 1;
}

const response: GeneratedType = {
  type: 1,
}

switch (response.type) {
  case DATA_TYPE.First:
  case DATA_TYPE.Second:
  default:
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/switch-exhaustiveness-check": [
      2,
      {
        "allowDefaultCaseForExhaustiveSwitch": true,
        "considerDefaultExhaustiveForUnions": false,
        "requireDefaultForNonUnion": false
      }
    ]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that eslint do not throw any error in switch

### Actual Result

ESLint throws error in switch:

```
Switch is not exhaustive. Cases not matched: 0 | 1
```

### Additional Info

I use `openapi-typescript` to generate response types based on swagger. It generates the following type:

```ts
type BackendV1Components = {
     schemas: {
        PublicationType: 0 | 1;
        PublicationViewModel: {
            type: BackendV1Components["schemas"]["PublicationType"];
        };
    }
}
```

I also have an enum that represent these values:

```ts
export enum PUBLICATION_TYPE {
  Standard = 0,
  LessonResult = 1,
}
```

But when I try to use the enum when processing the response, ESLint gives me the error described above:

```ts
const process = (model: BackendV1Components['schemas']['PublicationViewModel']) => {
  // Switch is not exhaustive. Cases not matched: 0 | 1
  switch (model.type) {
    case PUBLICATION_TYPE.Standard:
    case PUBLICATION_TYPE.LessonResult:
    default: 
  }
}
```

In addition, TypeScript correctly checks whether the enum member is equal to the case value. If I change the `LessonResult` member to `2`, then TypeScript will throw an error:

```ts
const process = (model: BackendV1Components['schemas']['PublicationViewModel']) => {
  // Switch is not exhaustive. Cases not matched: 0 | 1
  switch (model.type) {
    case PUBLICATION_TYPE.Standard:
    // Type 'PUBLICATION_TYPE.LessonResult' is not comparable to type '0 | 1'. ts(2678)
    case PUBLICATION_TYPE.LessonResult:
    default: 
  }
}
```

**Comment by kirkwaiblinger** [comment]
Hi @Afgan0r! This is an interesting case, since it relates to [no-unsafe-enum-comparison](https://typescript-eslint.io/rules/no-unsafe-enum-comparison/); from some points of view, enums aren't necessarily just their values. I think your case makes sense to consider a bug and fix, where the `case` type is an enum with a known value.

But, what about the reverse, where the enum is the discriminant and literals are the cases ([playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=types&fileType=.tsx&code=KYOwrgtgBAIgggFTgfQQTQAoFEoG8BQUUAYgJYBOAzgC5QC8UARAFrDkD2jANIVAMrAAxuxAATekwDyIYN3wBffPlFCANgENywKMJA0oo6gC5YiFOmwBuJZQDupaoIAWUABSGAlHl6D1lbSxsnEa8RABGWuoA1tZEvv5SMowhROGRMQpAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2HB2HQmkADRbtWbdkjD48APYcAIogBmo%2BPgDCw5IgBiN6ACiQqLiUogAylw8vGr4cIjG5hZkNkzIlAAmiNCOLgj4QSJiEtJ%2B0ACqTJSpqBgu8N6JSVDQiACOsJStua74ZQByqZXVBnVW3ibYAL6TALpNOAQk5FQ09MiMLKxMNrSwacJOiPRMsAC2CjZnxHqUyKlqkLr6kFozU0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))?

```ts
enum DATA_TYPE {
  First = "Zero",
  Second = "One",
}

declare const dt: DATA_TYPE;

switch (dt) { // Switch is not exhaustive. Cases not matched: DATA_TYPE.First | DATA_TYPE.Second
  case "Zero":
    break;
  case "One":
    break;
}
```

(note that the cases are also reported by no-unsafe-enum-comparison).


I guess we should consider this a bug too... Even if the typescript-eslint position is to prefer for `DATA_TYPE.First` to be used, that's a no-unsafe-enum-comparison concern, not switch-exhaustiveness-check.

@typescript-eslint/triage-team

**Comment by bradzacher** [comment]
If you inspect the type of `response` you'll see that TS also treats it the same way we do and it doesn't truly refine the code.

<img width="374" height="659" alt="screenshot of playground linked below" src="https://github.com/user-attachments/assets/61cf12dd-015b-4ab0-a271-7553fedacdbf" />

[playground](https://www.typescriptlang.org/play/?#code/KYOwrgtgBAIgggFTgfQQTQAoFEoG8BQUUAYgJYBOAzgC5QC8UADADSFQDKwAxgPYgAm9KAEZWAX3z5+3ADYBDcsCi8QNKDNLVg5OTIQBPAA7AAXHjbUjpplAA+IgNz4xTygHdNXABZQAFBq0dPSsAOktjAEpzIi45SiV4JFRMLBCyKmoTNiIA7V0DYzCrJwB6EqIKyqqoAD0AfmyoACNFOQBrJxi4hMQUdGwQzhV+LMrcoILgIuNS8ur52obKluB2zqhpADM5MBlMxvH80PDgWYX5+saVtedJKVkFJRU1UEhJswJKk7NEvpSnFz4dyeHy+V4QSbTYBRT7KbqwXrJAbpGijCrgyEnM7nIiXZatDpsWLxBFJfqpIZ8EaNDHHYr4Mo43FLCrXQlELY7PZooi0wpYhlzHF41kEgF3IA)

Note how the default branch is refined to `never` in the second switch but it is not refined in the first switch? That's because TS is not treating the default branch as unreachable due to refinements.

Enums are intended to be opaque -- they aren't intended to be used as transparent bags of values. So either direction seems correct to error on to me, and it matches TS's behaviour as well.

**Comment by Afgan0r** [comment]
Thanks for your reply @kirkwaiblinger. This makes sense, and `no-unsafe-enum-comparison` error would explain why the cases are not considered exhaustive. I agree that it's better to create a new issue for the bug in `no-unsafe-enum-comparison` rule.

Is there a way to create a new issue with without describing it from scratch, but simply by adding a link to this issue?

**Comment by kirkwaiblinger** [comment]
So, are we saying that we think the appropriate resolution is to add detection for switched-literal-union-with-enum-case (after the existing pattern for switched-enum-with-literal-union-case) to no-unsafe-enum-comparison rather than making any change to the switch-exhaustiveness-check behavior?

**Comment by kirkwaiblinger** [comment]
This might require a bit of a change to the existing logic for `no-unsafe-enum-comparison` outside of switch statements as well since it doesn't currently report a _union_ of literals being compared to an enum type; just an enum compared to a (single) literal. 

([playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=types&fileType=.tsx&code=KYOwrgtgBAcgqgWQEIFEBKB9F8FQN4BQUUAYgJYBOAzgC5QC8UADADRFQDKwAxgPYgATBlACMbAL4ECNAJ4AHYLEgAjYBTggy-YUygAfUQG4pAngBsAhhUV8QtKOAir1m-gC4lTtRq0hjp7ktrKFt7R2cAGTIaNQszDyZjKXDvVxAGekYcVExsRAA6cmoaQygAejLYAHkAFSg0FAAFKrQalAARAhSKKJiKOIysxBysHELKWlKKqGs5XgoYgSkCKgB3aO4ACygACm6ffgBKfHLK5RkoGk3FCjAzYAByKigzXgBzMm4WS82yZ9XeHcBCAHnRVDNgHMFsAhFQyCBuIpok8oBYoGA0vl2NwLFRFNl0KMCkVaG5sbj8cNCXkEPkuLYBGTiMRlNYLABrYySIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2HB2HQmkADRbtWbdkjD48APYcAIogBmo%2BPgDCw5IgBiN6ACiQqLiUogAylw8vGr4cIjG5hZkNkzIlAAmiNCOLgj4QSJiEtJ%2B0ACqTJSpqBgu8N6JSVDQiACOsJStua74ZQByqZXVBnVW3ibYAL6TALpNOAQk5FQ09MiMLKxMNrSwacJOiPRMsAC2CjZnxHqUyKlqkLr6kFozU0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
```ts
enum NUMBER_ENUM {
  First = 0,
  Second = 1,
}

type NumberUnion = 0 | 1;

declare const numberUnion: NumberUnion;
declare const numberLiteral: 0;


numberUnion === NUMBER_ENUM.First; // NOT REPORTED
numberLiteral === NUMBER_ENUM.First; // reported


switch (numberUnion) { // by the rule's logic, this wouldn't be reported since it's a union.
  case NUMBER_ENUM.First:
  case NUMBER_ENUM.Second:
    break;
}
```
